// angular-ui-router/src/transition/transitionService.js
/** @module transition */ /** for typedoc */
/// <reference path='../../typings/angularjs/angular.d.ts' />
var transition_1 = require("./transition");
var hookRegistry_1 = require("./hookRegistry");
/**
 * The default transition options.
 * Include this object when applying custom defaults:
 * let reloadOpts = { reload: true, notify: true }
 * let options = defaults(theirOpts, customDefaults, defaultOptions);
 */
exports.defaultTransOpts = {
    location: true,
    relative: null,
    inherit: false,
    notify: true,
    reload: false,
    custom: {},
    current: function () { return null; }
};
/**
 * This class provides services related to Transitions.
 *
 * Most importantly, it allows global Transition Hooks to be registered, and has a factory function
 * for creating new Transitions.
 */
var TransitionService = (function () {
    function TransitionService() {
        this._defaultErrorHandler = function $defaultErrorHandler($error$) {
            if ($error$ instanceof Error)
                console.log($error$);
        };
        this._reinit();
    }
    TransitionService.prototype.defaultErrorHandler = function (handler) {
        return this._defaultErrorHandler = handler || this._defaultErrorHandler;
    };
    TransitionService.prototype._reinit = function () {
        hookRegistry_1.HookRegistry.mixin(new hookRegistry_1.HookRegistry(), this);
    };
    /**
     * Creates a new [[Transition]] object
     *
     * This is a factory function for creating new Transition objects.
     *
     * @param fromPath
     * @param targetState
     * @returns {Transition}
     */
    TransitionService.prototype.create = function (fromPath, targetState) {
        return new transition_1.Transition(fromPath, targetState);
    };
    return TransitionService;
})();
// todo: move to Router class
exports.$transitions = new TransitionService();
$TransitionProvider.prototype = exports.$transitions;
function $TransitionProvider() {
    this._reinit.bind(exports.$transitions)();
    /**
     * @ngdoc service
     * @name ui.router.state.$transitions
     *
     * @description
     * The `$transitions` service is a registry for global transition hooks, and is a factory for Transition objects.
     */
    this.$get = function $get() {
        return exports.$transitions;
    };
}
exports.$transitionsProvider = $TransitionProvider;
angular.module('ui.router.state')
    .provider('$transitions', exports.$transitionsProvider);
